Centralize and Fix DOMPurify Sanitizer

Zachary-Squires opened the pull request:
# Description

Merges changes from PR #1488 (https://github.com/OpenEnergyDashboard/OED/pull/1488) and fixes the issue where the update to DOMPurify and jsdom was not reflected in the package-lock.json file. Now that DOMPurify is properly implemented the cross-site scripting vulnerability when users upload meter data should be remediated. 

Developed and Implemented by:
Zachary Squires - https://github.com/Zachary-Squires 
Zachary Bates - https://github.com/Zach-O-Bates
Krista Clanton - https://github.com/Quillen09

## Type of change

- [ ] Note merging this changes the database configuration.
- [ ] This change requires a documentation update

## Checklist

(Note what you have done by placing an "x" instead of the space in the [ ] so it becomes [x]. It is hoped you do all of them.)

- [x ] I have followed the [OED pull request](https://openenergydashboard.org/developer/pr/) ideas
- [ ] I have removed text in ( ) from the issue request
- [x ] You acknowledge that every person contributing to this work has signed the [OED Contributing License Agreement](https://openenergydashboard.org/developer/cla/) and each author is listed in the Description section.

## Limitations

N/A

Landon-Wivell committed:
feat(security): centralize DOMPurify sanitizer and patch XSS in success/failure responses

Landon-Wivell committed:
Adding the commit from 1488 called "fix: resolved PR feedback on sanitizer imports and tests"

Landon-Wivell committed:
fix: resolved PR feedback on sanitizer imports and tests

Zachary-Squires committed:
Create README.md

Zachary-Squires committed:
Create thing-in-docs-folder

squid game

Andrew-Bonner committed:
add a file for team meeting schedule

Zachary-Squires committed:
Update thing-in-docs-folder

Zachary-Squires committed:
Merge branch 'main' of https://github.com/CS495-Fall2025/Team-5

Zachary-Squires committed:
Added the commit from 1488 called "feat(security): centralize DOMPurify sanitizer and patch XSS in success/failure responses"

Zachary-Squires committed:
Ran npm install, should bring jsdom and dompurify
up to date. package-lock.json and package.json
should match now.

Zachary-Squires committed:
Restore executable permissions on scripts and binaries

Zachary-Squires committed:
Merge pull request #1 from Andrew-Bonner/team_b

Cross Site Scripting fix.

Zachary-Squires committed:
Normalize all line endings to LF

Zachary-Squires committed:
Merge pull request #2 from Andrew-Bonner/team_b

Normalize all line endings to LF

huss commented on pull_request_template.md:
```diff
@@ -23,4 +23,4 @@ Fixes #[issue]
 ## Limitations
 
-(Describe any issues that remain or work that should still be done.)
+(Describe any issues that remain or work that should still be done.)
```
I suggest the blank line at the end of this file be added back so no changes are shown (this is the only one).

huss commented on .gitattributes:
```diff
@@ -14,4 +14,8 @@
-*.RTF	 diff=astextplain
\ No newline at end of file
+*.RTF	 diff=astextplain
+*.sh text eol=lf
```
The change of CRLF vs LF has been an issue for others and caused confusion. Thus, I'm okay with getting git to normalize the ending to LF. Having said that, I wonder if this change should be in a separate PR that involves that change. I also think (at least in the past) that several files have CRLF and these could be fixed (via easy commands) so the entire repo is LF as part of doing this. What do you think?

huss commented on package.json:
```diff
@@ -67,12 +67,14 @@
 		"csv": "~5.3.2",
 		"csv-stringify": "~5.6.5",
 		"d3": "~7.8.5",
+		"dompurify": "~3.2.6",
```
dompurify now has a 3.3.0 version so maybe use that (verify works fine). jsdom is at 27.0.1.

huss replied:
dompurify was updated but not jsdom.

huss replied:
Testing found that these versions are not compatible with our current version of node. Thus, I switched them back to the slightly earlier version that works fine.

huss commented on src/server/test/something.csv:
```diff
@@ -0,0 +1 @@
+1,2,3
```
This seems the same as src/server/test/crossSite/something.csv. The other file is used but I could not find where this is used. Is it needed?

huss commented on src/server/services/utils/sanitizer.js:
```diff
@@ -0,0 +1,8 @@
+const { JSDOM } = require('jsdom');
```
This should have a copyright notice at the top.

huss commented on src/server/test/crossSite/crossSite.js:
```diff
@@ -0,0 +1,28 @@
+const { chai, mocha, expect, app, testUser } = require('../common');
+
+mocha.describe('Cross site', () => {
+	mocha.it('test 1: tests for sanitization of HTML', async () => {
```
I propose to remove "test1: " from the mocha.it. First, OED does not generally do that. Second, there is only one test.

huss commented on src/server/test/crossSite/crossSite.js:
```diff
@@ -0,0 +1,28 @@
+			.field('meterName','<img src=x onerror="alert(document.domain)">')
+			.field('gzip', "no")
+			.attach('csvfile', 'src/server/test/crossSite/something.csv');
+        expect(res.text).to.include('<img src="x">');
```
First, the indenting is spaces not tabs. Second, this tests the tag was modified as expected which is good. I'm wondering if it would be a stronger test to see if "alert" or "document" is not present. What do you think?

huss replied:
I also just noticed that the formatting is off. VSC document format can fix that.

huss commented on src/server/test/crossSite/crossSite.js:
```diff
@@ -0,0 +1,28 @@
+			.field('gzip', "no")
+			.attach('csvfile', 'src/server/test/crossSite/something.csv');
+        expect(res.text).to.include('<img src="x">');
+		expect(res).to.have.status(400); 
```
Normally OED checks the status first since if it is wrong then other tests are likely to fail.

huss commented on src/server/services/csvPipeline/success.js:
```diff
@@ -2,35 +2,33 @@
+    const safeComment = DOMPurify.sanitize(comment);
+    res.status(200).send(`<h1>SUCCESS</h1>${safeComment}`);
 }
 
```
Please remove the second blank line between the functions.

huss commented on src/server/services/csvPipeline/success.js:
```diff
@@ -2,35 +2,33 @@
 }
-
-module.exports = { success, failure };
+module.exports = { success, failure };
```
Please add back the blank line after the function but before module.exports.

huss commented on src/server/services/csvPipeline/success.js:
```diff
@@ -2,35 +2,33 @@
 function success(req, res, comment = '') {
-	res.status(200) // 200 OK
-	.send(`<h1>SUCCESS</h1>${comment}`);
+    const safeComment = DOMPurify.sanitize(comment);
```
All indenting should be tabs. VSC has a convert indentation to tabs for this.

huss commented on src/server/routes/response.js:
```diff
@@ -2,32 +2,31 @@
-	res.status(200) // 200 OK
-	.send(comment);
+function success(res, comment = '') {
+    const safeComment = DOMPurify.sanitize(comment);
```
Intent via tabs per other comment.

huss reviewed (changes requested):
Thanks to @Zachary-Squires, @Zach-O-Bates & @Quillen09 for their first contribution to OED. I'm sorry this took so long as my time was directed elsewhere. I have made a some comments to consider. I note that several are not your changes to the previous PR but should be addressed. Second, here are some file-level comments:

- ~Several files seem to have the only change as the file attributes of 755 (rwxr-xr-x) TO 644 (rw-r--r--). These are:~
  - ~src/scripts/. These were set the have execute (x) so that copying them to the cron location (per site installation directions) would make them executable as normally done. It saves doing this manually. I see that one could accidentally execute the file via the script directory and this change avoids that. However, I think it might be better to either do this for all cron sample files (the refresh*.bash do not seem changed) or none. It also seems a different area than the pull request and is not in the description. I'm open to a discussion of this but would be inclined to make this a separate pull request if people want to do this. The files that seem to have these changes are:~
    - ~sendLogEmailCron.bash~
    - ~updateEgaugeMetersOEDCron.bash~
    - ~updateMamacMetersOEDCron.bash~

  - ~docker-compose.yml changed but I think that is okay as it is a configuration file.~

- ~Five files in src/server/test/web/readingsData/*.csv are shown as having changes but it seems to show the line as deleted and then the same line added. I'm not sure why this is happening but maybe it relates to the change in .gitattributes? Do you have any ideas?~
  -  Update: I manually checked each file and there are no diffs. I suspect this is due to saving a file (maybe due to line endings that were then put back?) and this is confusing git. I'm going to ignore this listed change and accept it. 
    - This was resolved in PR #1591 that is replacing this PR so it is done.

Please let me know if you have any questions/thoughts.

Zachary-Squires committed:
Change to not allow default postgres or secret token in OED production.

Changed the installOED.sh file to generate randomized values for OED_TOKEN_SECRET and POSTGRES_PASSWORD, which are then stored in .env file. Changed docker.compose.yml to draw from the .env file if variables exist there. When in dev mode with the default values the user will be warned in the console when OED is started. Also implemented a warning for the mailing variables, if the method is set and a default remains in the rest the user will be warned in production.

Zachary-Squires committed:
Merge branch 'OpenEnergyDashboard:development' into team_b

Zachary-Squires committed:
Merge branch 'team_b' of https://github.com/Andrew-Bonner/OED into team_b

Zachary-Squires committed:
Merge pull request #3 from Andrew-Bonner/team_b

Team b

Zach-O-Bates committed:
Fix PR suggestions

Zachary-Squires committed:
Revert "Merge pull request #3 from Andrew-Bonner/team_b"

This reverts commit 489cc17dad9f26f461e3764c29e8845632f2fc6b, reversing
changes made to f38a83effa8259a03708bc6efd35c8c01afd9730.

Zach-O-Bates committed:
Merge branch 'team_b' into development

huss commented on src/server/test/crossSite/crossSite.js:
```diff
@@ -0,0 +1,34 @@
+			process.exit(1)
+		}
+
+		if (!res.text.includes('<iimg src="x>')){
```
Should this be ``iimg`` or img?

Could you help me understand the two if statements? Is there any chance that no test is done? Maybe the test can be simplified so it always returns the text to test (we can discuss if that would help).

huss commented:
@Zachary-Squires, @Zach-O-Bates & @Quillen09 I looked over the changes. I've left the comments that did not seem to be addressed and added a couple of more comments. I'm waiting to do final testing until everything is resolved. Please let me know if you have any questions/thoughts.

Zach-O-Bates committed:
Simplify test to better align with mentor suggestions

Zachary-Squires committed:
Update .gitattributes

Remove line endings fix, will be implemented separately.

Zachary-Squires committed:
Update docker-compose.yml

Restoring default value of OED_PRODUCTION

Zach-O-Bates committed:
Update jsdom package

huss commented on src/server/test/crossSite/crossSite.js:
```diff
@@ -0,0 +1,28 @@
+			.attach('csvfile', 'src/server/test/crossSite/something.csv');
+		expect(res).to.have.status(400);
+		expect(res.text).to.include('<img src="x">');
+	})
```
I think testing for .not having "alert" (and maybe other strings?) would strengthen this test.

huss commented on src/client/app/components/unit/EditUnitModalComponent.tsx:
```diff
@@ -410,7 +410,7 @@ export default function EditUnitModalComponent(props: EditUnitModalComponentProp
 										type='text'
 										autoComplete='on'
-										onChange={e => { handleStringChange(e); }}
+										onChange={e => {handleStringChange(e);}}
```
The changes in this file all appear to be formatting. The format changes are not what VSC does by default for js file. Given this, I think they all should be removed.

huss commented on src/client/app/redux/selectors/barChartSelectors.ts:
```diff
@@ -12,15 +12,13 @@ import getGraphColor from '../../utils/getGraphColor';
 import { createAppSelector } from './selectors';
 import { selectAreaScalingFromEntity, selectNameFromEntity } from './entitySelectors';
 import { selectPlotlyMeterDeps, selectPlotlyGroupDeps } from './plotlyDataSelectors';
```
I'm wondering if this change about selectSelectedLanguage is showing since that change not yet merged into this PR or something else? Either way, it is a needed change so figuring out would be good.

huss replied:
I'm noting that this is resolved but it did change the git record to show it as a recent change. The work was originally done in PR #1547.

huss commented on docker-compose.yml:
```diff
@@ -111,4 +111,4 @@ services:
         Xvfb :99 -screen 0 1024x768x16"
-        
\ No newline at end of file
+        
```
Could this new, blank line (below) be removed so the file does not should changes?

huss commented on .gitattributes:
```diff
@@ -14,4 +14,4 @@
 *.rtf	 diff=astextplain
-*.RTF	 diff=astextplain
\ No newline at end of file
+*.RTF	 diff=astextplain
```
Same new, blank line as in the docker-compose.yml file.

Zachary-Squires replied:
It's the exact same as the gitattributes in OED, I'm honestly not sure why it's still showing as a change

huss replied:
I merged development, resolved the conflict, pushed and this fixed the issue.

huss replied:
I resolved via a merge.

huss reviewed (changes requested):
Thanks to @Zachary-Squires, @Zach-O-Bates & @Quillen09 for the updated code. I've made a few comments to address. Also, The comment from 10/27 on the scripts and .csv files showing changes is still present. I think it needs to be resolved. Please let me know if something is not clear or you need anything.

Zachary-Squires committed:
Addressing Suggestions for PR 1544

Zachary-Squires committed:
Mark cron scripts as executable

Zachary-Squires committed:
Mark docker-compose.yml as executable

Zachary-Squires committed:
Update docker-compose.yml

huss committed:
try to resolve some blank line change issues

huss committed:
Merge remote-tracking branch 'origin/development' into pr/Zachary-Squires/1544

huss committed:
revert dompurify & jsdom versions

The ones suggested in my comment are not compatible
with the current OED version of node so put them
back to a slightly older version.

huss committed:
fix so check header ignores new CSV file

huss reviewed (changes requested):
@Zachary-Squires, @Zach-O-Bates & @Quillen09 I looked over the revised code. I also made some commits you should review to make sure you think they are okay. I marked comments that are done. There is still one comment that should be addressed but otherwise this looks goodl

Zach-O-Bates committed:
Add meeting times to team_schedule

Zach-O-Bates committed:
Add final presentation date to team schedule

Zach-O-Bates committed:
Enhance README with project details and contributors

Added project overview and contributors list.

Zach-O-Bates committed:
Adjust file structure according to the Github folder document

Zach-O-Bates committed:
Delete .DS_Store

Zach-O-Bates committed:
Create the team and client folders for meeting notes

Zach-O-Bates committed:
Merge branch 'main' of https://github.com/CS495-Fall2025/Team-5

Zach-O-Bates committed:
Summarize and organize first client meeting notes

Zach-O-Bates committed:
Add notes from client meeting on September 17, 2025

Zach-O-Bates committed:
Rename 9-17-2025-ClientMeeting to 9-17-2025-ClientMeeting.md

Zach-O-Bates committed:
Rename client meeting document for consistency

Zach-O-Bates committed:
Add notes from the client meeting on September 22, 2025

Zach-O-Bates committed:
Update and rename team_schedule to team_schedule.md

Zach-O-Bates committed:
Fix typo in presentation date heading

Zach-O-Bates committed:
Add meeting notes for client meeting on 10-01-2025

Zach-O-Bates committed:
Removed meeting 10-01 from a mistakenly created folder

Zach-O-Bates committed:
Add client meeting notes for October 8, 2025

Documented the agenda and discussions from the client meeting held on October 8, 2025, including group tasks and options for Issue 1 and Issue 2 solutions.

Zach-O-Bates committed:
Add notes for October 15, 2025 Client Meeting

Added notes regarding mentor's availability for the meeting.

Zach-O-Bates committed:
Adjust folder structure in client meetings

Zach-O-Bates committed:
Added test for the second route in crossSite.js

Andrew-Bonner committed:
updated the expected to be a status code 401

Andrew-Bonner committed:
Add files via upload

Andrew-Bonner committed:
Create 10-22-2025-ClientMeeting.md

Zach-O-Bates committed:
Create 10-29-2025-ClientMeeting.md

Zach-O-Bates committed:
Create 11-5-2025-ClientMeeting.md

Zach-O-Bates committed:
Create 11-12-2025-ClientMeeting.md

Zach-O-Bates committed:
Create 11-19-2025-ClientMeeting.md

Zach-O-Bates committed:
Create 11-26-2025-ClientMeeting.md

Zach-O-Bates committed:
Create 12-3-2025-ClientMeeting.md

Andrew-Bonner committed:
Update 10-22-2025-ClientMeeting.md

Andrew-Bonner committed:
Update 10-29-2025-ClientMeeting.md

Andrew-Bonner committed:
Update 11-12-2025-ClientMeeting.md

Zach-O-Bates committed:
Add notes for 11-5-2025 Client Meeting

Added meeting notes for the client meeting held on November 5, 2025, including agenda items and group progress updates.

Zach-O-Bates committed:
Revise client meeting notes for 11-5-2025

Updated meeting notes with group progress and mentor suggestions.

Zach-O-Bates committed:
Add notes for client meeting on November 12, 2025

Added meeting notes for the client meeting on November 12th, 2025.

Zach-O-Bates committed:
Add meeting notes for client meeting on 12-3-2025

Added meeting notes for client meeting on November 12th, 2025, detailing issues discussed and progress made.

Andrew-Bonner committed:
Update 11-19-2025-ClientMeeting.md

Zach-O-Bates committed:
Merge branch 'OpenEnergyDashboard:development' into development

Oykunle committed:
Create Meeting-Minutes folder for weekly notes

Oykunle committed:
weekly meeting notes (Weeks 1–9)

BrianRaymond800 committed:
Create risk-log.md

BrianRaymond800 committed:
Update risk-log.md

Added each issue as a risk, with probability, impact, and solution.

Zach-O-Bates committed:
Move meeting minutes into the team meeting folder and add UML diagrams

Zach-O-Bates committed:
Move risk log into the documents folder

Zach-O-Bates committed:
Add communication plan document

huss commented on src/server/test/crossSite/crossSite.js:
```diff
@@ -0,0 +1,71 @@
+		expect(res).to.have.status(400);
+		expect(res.text).to.include('<img src="x">');
+
+		const xssIndicators = [
```
One small suggestion. ``xssIndicators`` is duplicated below so how about putting it once above the ``mocha.it`` and after the ``mocha.describe`` so it can be reused but still scoped to these tests.

huss reviewed (changes requested):
Thanks to @Zach-O-Bates for the updates that addressed the remaining comment. I have made one small comment to consider and then I think this is ready to merge.

Oykunle committed:
Add meeting notes for 09/01/2026

Oykunle committed:
Add meeting notes for 09/01/2026

Zach-O-Bates committed:
Remove the multiple instances of xss indicators
Add a single instance before the tests

Zach-O-Bates committed:
Merge remote-tracking branch 'team5/main' into development

huss commented on src/server/test/crossSite/crossSite.js:
```diff
@@ -0,0 +1,59 @@
+/* Run in OED Docker web container terminal/shell:
+npm run testsome src/server/test/crossSite/crossSite.js */
+const { chai, mocha, expect, app, testUser } = require('../common');
+const xssIndicators = [
```
Thanks for changing so only one version. How about moving it right below the mocha.describe so it isn't globally scoped?

huss replied:
Resolved in replacement PR #1591.

huss reviewed (changes requested):
@Zachary-Squires Thanks for the updated coded. I made some small comment to consider. Otherwise, testing & review found everything looks good.

Zach-O-Bates committed:
Revert "Merge remote-tracking branch 'team5/main' into development"

This reverts commit 45ff19ecd40ae4321c15005ce0d7d2befffd89ba, reversing
changes made to 73b39d814c2525024ed90de85c89cc15ce38fec6.

Andrew-Bonner committed:
Merge branch 'OpenEnergyDashboard:development' into development

huss commented:
All comments have either been resolved here or in PR #1591 that supersedes this one. Thus, this PR is being closed.
